draft-ietf-rats-architecture.md

```diff
@@ -360,6 +363,42 @@ are signed.  For example, values might have been generated at boot, and then
 used in claims as long as the signer can guarantee that they cannot have changed
 since boot.
 
+The term freshness is applied to both Evidence and Attestation Results in
+this document. Unfortunately, the term freshness is also a bit overloaded and
+in order to illustrate the application of the term in this document better,
+two characteristics that can be part of freshness are elaborated on below:
+recentness and uniqueness:
+
+Recentness:
+
+: a quality of Evidence that is based on the time that has passed between
+creation of the evidence by an Attester (typically by "sealing" a Claims set
+via signing) and the reception of that Evidence by a Verifier or (relaying
+party).
+
+: Evidence can be not recent enough, if the Evidence created from collected
```

Comment:
Unless I'm missing something big, this sentence looks like a truism to me?
I think this should state what defines the "too old" criterion.

Reply:
While I understand your point of view.... I literally strongly disagree. It very much can be used specifically for recentness in the context of CHARRA.

Reply:
The thing I wanted to say is that any concept of "recentness" must come with a (explicit or implicit) "max-age" label, otherwise it isn't actionable.

Reply:
Example: a nonce provided by a Verifier in a network protocol to retrieve Evidence from an Attester that is subsequently used in creating the Evidence enables a Verifier to assert that the reply from the Attester including the evidence was created upon the just emitted request of the Verifier; therefore establishing the most common method of providing a proof-of-recentness via a challenge/response remote attestation (CHARRA).

Maybe I am missing something obvious here, but how is this mechanic not addressing the appraisal of "recent creation of evidence"?

Reply:
It's the _uniqueness_ property of the nonce that guarantees that the Evidence is fresh.  A nonce bears no recentness.

Reply:
Then no value, including the timestamp of a certificate, bears no inherent recentness.

But that is not correct in the context of the Challenge/Response handshake. Again, maybe I am missing something very obvious. Could you provide me with a protocol-based exemplary use of a nonce, generated by the challenger, where the reply that uses the nonce (most likely in a "sealed" way, such as digital signing) does not bear a proof-of-recentness by using that nonce in its reply?

Reply:
An additional clarifying comment:

If you add a timestamp that was generated on a responder in evidence, the "meaning of recentness it bears" is based on a trust relationship created by an endorser that vouches for the fact that "timestamps from that attestation Evidence provenance" are to be trusted. Alternatively, a challenger can simply establish a similar bi-lateral trust relationship by conducting a Challenge/Response handshake here. It is a smaller scoped trust relationship, it is temporary, and the challenger must at least be RESTful in order not to forget that it created the nonce in the Challenge. But then - in this context - the nonce created by the challenger "bears recentness" from the point of view of the Challenger, I would think.

Reply:
The max-age in this case is not included in the evidence as a value, it is simply state (like... a Verifier measuring a time interval to the best of its capability) maintained by the challenger. In that way, like a secure channel enabling you to omit a signature, the nonce enables you to omit a max-age while retaining being actionable.

Reply:
> Then no value, including the timestamp of a certificate, bears inherent recentness characteristics, yes?

Not by itself.  You also need a synchronised clock and an expiry threshold to determine “freshness by recentness”, at least in the sense I intend it.

> But that is not correct in the context of the Challenge/Response handshake. Again, maybe I am missing something very obvious. Could you provide me with a protocol-based exemplary use of a nonce, generated by the challenger, where the reply that uses the nonce (most likely in a "sealed" way, such as digital signing) does not bear a proof-of-recentness by using that nonce in its reply?

This is IMO “freshness by uniqueness”, in fact it doesn’t require a clock nor an expiry to work, just a non-repeating source of byte strings.

Reply:
I guess my main point here is the following: if we need to go into the semantic finesse of distinguishing between the recentness and uniqueness aspects of freshness, then we need to make it worth the words we spend by _clearly_ separating the two.

Since I've diluted it across a few different comments, it's maybe worth summing up the way I see the distinction working in practice :
- uniqueness needs a source of non-repeatable byte strings and is completely driven by the requestor;
- recentness needs synchronised clocks between requestor and responder and is conveyed via a timestamp plus an expiration threshold associated with the provided evidence / attestation result.

The problem I have with the current text when it says that nonces convey recentness is it seems to be blurring that distinction, therefore voiding the overall "semantic finesse" plan :-)

Reply:
> > Then no value, including the timestamp of a certificate, bears inherent recentness characteristics, yes?
> 
> Not by itself. You also need a synchronised clock and an expiry threshold to determine “freshness by recentness”, at least in the sense I intend it.

Then I agree that without context, values never bear a meaning of recentness.



> > Then no value, including the timestamp of a certificate, bears inherent recentness characteristics, yes?
> 
> Not by itself. You also need a synchronised clock and an expiry threshold to determine “freshness by recentness”, at least in the sense I intend it.
> 
> > But that is not correct in the context of the Challenge/Response handshake. Again, maybe I am missing something very obvious. Could you provide me with a protocol-based exemplary use of a nonce, generated by the challenger, where the reply that uses the nonce (most likely in a "sealed" way, such as digital signing) does not bear a proof-of-recentness by using that nonce in its reply?
> 
> This is IMO “freshness by uniqueness”, in fact it doesn’t require a clock nor an expiry to work, just a non-repeating source of byte strings.

Well, the Verifier has to retain state, "remembering the nonce" until a certain time-out, which I think most certainly requires a clock. Otherwise, networks protocols would have a hard time.

The curious mixture of "freshness by uniqueness" is exactly the reason why I wrote the text. Typically "freshness" is used without context here. And we decided to provide more context because it is not "freshness by uniqueness" but "recentness & uniqueness" called freshness for simplicity, I think.

Reply:
> I guess my main point here is the following: if we need to go into the semantic finesse of distinguishing between the recentness and uniqueness aspects of freshness, then we need to make it worth the words we spend by _clearly_ separating the two.

Absolutely agreed.

> * uniqueness needs a source of non-repeatable byte strings and is completely driven by the requestor;
> 
> * recentness needs synchronised clocks between requestor and responder and is conveyed via a timestamp plus an expiration threshold associated with the provided evidence / attestation result.

I am not entirely sure that uniqueness is always completely driven by the challenger/requester, but it most certainly can and is.

I agree that recentness is based on synchronization and timestamps in TUDA, of course. I still do not understand how that is required for a Challenge/Response handshake, because there are conveyance mechanisms today that only need a clock on the Challenger side, measuring the time between a Challenge including a nonce and a Response including a digest of the nonce (in a secure channel or digitally signed).

 

> The problem I have with the current text when it says that nonces convey recentness is it seems to be blurring that distinction, therefore voiding the overall "semantic finesse" plan :-)

You are correct that a nonce value by itself does not convey recentness :-) We wildly agree that even timestamp values do not do that. I think that is typically way harder to agree on. A nonce in a Challenge/Response interaction does convey recentness, but due to the context of the protocol mechanism - not by the value itself. I tried to always provide the context of the interaction model in the text, but i seems I was not successful with that.

Reply:
> Well, the Verifier has to retain state, "remembering the nonce" until a certain time-out, which I think most certainly requires a clock. Otherwise, networks protocols would have a hard time.

I disagree, the existence of a garbage collector that sweeps allocated nonces at regular intervals (i.e., the "remembering until time-out") is just an implementation detail, it's not fundamental.

What a nonce gives you is the assurance that *when* the transaction completes -- maybe one ms after the end of the universe! ;-) -- the response you get back is uniquely bound to your original request and there is no other transaction that could have produced the same response.  This is what I mean by freshness by uniqueness.

Reply:
Okay :-) Let me rephrase that. A Challenger that intends to retrieve Evidence in a finite, has a clock and that intends to obtain a notion of recentness can track a time time interval only using its clock to do this. The indicator for the duration here (duration being a time interval where the left time is an epoch starting with 0 at which the Challenge including a nonce is emitted and at the right time evidence including a digest of a nonce is received) is the time that has passed between Challenge and Response.

Very small RESTful interactions as enabled via CoAP enable this (even with using an observe option) notion of recentness - intrinsically. Retaining state for an operation is not an implementation detail, I think. It is, for example required to associate CoAP tokens. Typically protocols use time-outs to avoid waiting until the end of the universe - and I think that is not an implementation detail, too, but an interaction model capability. The binding to what you call an transaction here, which typically is a operation I think, is what is the essential part to get a notion of recentness. 

In contrast to your point, my point is that a re-use of a nonce that is cryptographically (and in this case uniquely) "bound" to a Response does not require synchronized clocks due to the "recentness" based on the uniqueness that is provided by the "arbitrary number used only once in a cryptographic communication" (nonce) intrinsically.

In consequence, I think:

1.) "recentness by uniqueness" is an over-simplification (you could receive two responses based on the same nonce, violating the concept of how to use a nonce, and still only process the most recent response via sequence or clock use), and
2.) "recentness needs synchronized clocks between requestor and responder and is conveyed via a timestamp plus an expiration threshold associated with the provided evidence / attestation result." is not correct, due to being too specific (you do not always need synchronized clocks) and unnecessary limiting (you do not necessarily need time-stamps and or expiration dates).

If I would only look at the interaction model of TUDA, I would agree! :-) But looking at charra, tuda and subscribed evidence, I think I still cannot agree.

Reply:
OK, it seems like we agree to disagree then :-)

Reply:
I assume there is common ground somewhere :-) Illustrating the different aspects about freshness for conceptual messages generally, might help us to converge maybe? Appraisal Policies enable a period or maybe even duration of validity for Attestation Evidence. Then synchronized clocks are a vital part here. 

Also, there might be other opinions out there?

Reply:
> I assume there is common ground somewhere :-) 

As much as I'd like to, I think our paths diverge a bit too early to try and reconcile.

Re-reading your last batch of replies, though, I've found something I think you are 100% correct:

> you do not always need synchronized clocks

true!  You only need that when you the requestor depends on an _explicit_ timestamp set by the responder.